src/include/duckdb/common/types/timestamp.hpp

```diff
@@ -67,21 +67,49 @@ struct timestamp_t { // NOLINT
 	timestamp_t &operator-=(const int64_t &delta);
 
 	// special values
-	static timestamp_t infinity() { // NOLINT
+	static constexpr timestamp_t infinity() { // NOLINT
 		return timestamp_t(NumericLimits<int64_t>::Maximum());
-	}                                // NOLINT
-	static timestamp_t ninfinity() { // NOLINT
+	}                                          // NOLINT
+	static constexpr timestamp_t ninfinity() { // NOLINT
 		return timestamp_t(-NumericLimits<int64_t>::Maximum());
-	}                                   // NOLINT
-	static inline timestamp_t epoch() { // NOLINT
+	}                                             // NOLINT
+	static constexpr inline timestamp_t epoch() { // NOLINT
 		return timestamp_t(0);
 	} // NOLINT
 };
 
-struct timestamp_tz_t : public timestamp_t {};  // NOLINT
-struct timestamp_ns_t : public timestamp_t {};  // NOLINT
-struct timestamp_ms_t : public timestamp_t {};  // NOLINT
-struct timestamp_sec_t : public timestamp_t {}; // NOLINT
+struct timestamp_tz_t : public timestamp_t { // NOLINT
+	static constexpr timestamp_t infinity() {
+		return timestamp_t::infinity();
+	}
+	static constexpr timestamp_t ninfinity() {
+		return timestamp_t::ninfinity();
+	}
+};
+struct timestamp_ns_t : public timestamp_t { // NOLINT
+	static constexpr timestamp_t infinity() {
+		return timestamp_t::infinity();
+	}
+	static constexpr timestamp_t ninfinity() {
+		return timestamp_t::ninfinity();
+	}
+};
+struct timestamp_ms_t : public timestamp_t { // NOLINT
+	static constexpr timestamp_t infinity() {
+		return timestamp_t(int64_t(NumericLimits<int64_t>::Maximum() / 1000) * 1000);
```

Comment:
Let's always use `timestamp_t::infinity`.

@hawkfish could you perhaps have a look at this, it looks like there is an issue when casting infinity to non-microsecond timestamp types, e.g.:

```sql
D select 'infinity'::timestamp_ms as ts;
┌───────────────────────────┐
│            ts             │
│       timestamp_ms        │
├───────────────────────────┤
│ 294247-01-10 04:00:54.775 │
└───────────────────────────┘
```